tests/test_missing.py

```diff
@@ -6,63 +6,278 @@
 import tempfile
 import unittest
 
+from netCDF4 import Dataset
+
 from activestorage.active import Active
 from activestorage.config import *
 from activestorage import dummy_data as dd
 
 import utils
 
 
-def _doit(testfile):
-    """ 
-    Compare and contrast vanilla mean with actual means
-    """
-    uri = utils.write_to_storage(testfile)
-    active = Active(uri, "data", utils.get_storage_type())
+def load_dataset(testfile):
+    """Load data as netCDF4.Dataset."""
+    ds = Dataset(testfile)
+    actual_data = ds["data"][:]
+    ds.close()
+
+    return actual_data
+
+
+def active_zero(testfile):
+    """Run Active with no active storage (version=0)."""
+    active = Active(testfile, "data", utils.get_storage_type())
     active._version = 0
     d = active[0:2, 4:6, 7:9]
-    mean_result = np.mean(d)
 
-    active = Active(uri, "data", utils.get_storage_type())
+    return d
+
+
+def active_two(testfile, valid_min=None, valid_max=None):
+    """Run Active with active storage (version=2)."""
+    active = Active(testfile, "data", utils.get_storage_type(),
+                    valid_min=valid_min, valid_max=valid_max)
     active._version = 2
     active.method = "mean"
     active.components = True
     result2 = active[0:2, 4:6, 7:9]
-    np.testing.assert_array_equal(mean_result, result2["sum"]/result2["n"])
+    active_mean = result2["sum"] / result2["n"]
+
+    return active_mean
 
 
-@pytest.mark.skipif(USE_S3, reason="Missing data not supported in S3 yet")
 def test_partially_missing_data(tmp_path):
     testfile = str(tmp_path / 'test_partially_missing_data.nc')
     r = dd.make_partially_missing_ncdata(testfile)
-    _doit(testfile)
 
-@pytest.mark.skipif(USE_S3, reason="Missing data not supported in S3 yet")
+    # retrieve the actual numpy-ed result
+    actual_data = load_dataset(testfile)
+    unmasked_numpy_mean = np.mean(actual_data[0:2, 4:6, 7:9])
+    print("Numpy masked result (mean)", unmasked_numpy_mean)
+
+    # write file to storage
+    testfile = utils.write_to_storage(testfile)
+
+    # NOT numpy masked to check for correct Active behaviour
+    no_active_mean = np.mean(active_zero(testfile))
+    print("No active storage result (mean)", no_active_mean)
+
+    active_mean = active_two(testfile)
+    print("Active storage result (mean)", active_mean)
+
+    if not USE_S3:
+        np.testing.assert_array_equal(unmasked_numpy_mean, active_mean)
+    else:
+        np.testing.assert_raises(AssertionError,
+                                 np.testing.assert_array_equal,
+                                 unmasked_numpy_mean, active_mean)
+    np.testing.assert_array_equal(no_active_mean, active_mean)
```

Comment:
need to open an issue about h5netcdf not constructing masks for the `data` member - it's annoying, but a great find from you - I am so set in netCDF4py mentality I didn't even think this could be a possibility :man_facepalming: